Serialization.md

```diff
@@ -0,0 +1,184 @@
+
+# MapEnumerable(...)
+
+If the class property that you need to serialize is an IEnumerable (a list, a OSPSuite Cache collection,), and there exists a serializer for the type of objects stored in th IEnumerable, you need to define the serialization of that property with the MapEnumerable() function, where you pass the enumerable and an method used to add objects to the defined IEnumerable. As an example, with an readonly list:
+
+```
+public class NewClass
+{
+   private readonly List<object> _allData;
+
+   public IReadOnlyList<object> AllData
+   {
+      get { return _allData; }
+   }
+
+   public void Add(object singleData)
+   {
+      _allData.Add(singleData);
+   }
+}
+
+.
+.
+.
+public class NewClassSerializer : OSPSuiteXmlSerializer<NewClass>
+{
+   public override void PerformMapping()
+   {
+      MapEnumerable(x => x.AllData, x => x.Add);
+   }
+}
+
+```
+
+# MapReference(...)
+
+Sometimes we have a class that has a reference to another object. Take the class `WeightedObservedData` in OSPSuite.Core for example:
+
+```
+   public class WeightedObservedData
+   {
+      public virtual DataRepository ObservedData { get; }
+      .
+      .
+      .
+      public WeightedObservedData(DataRepository observedData)
+      {
+         ObservedData = observedData;
+         .
+         .
+      }
+      .
+      .
+   }
+```
+
+In such a case the `Datarepository` that is `ObservedData` also exists in the Project and gets serialized and deserialized separately. We would not like to keep multiple copies of the same object in our project file, and therefore what we are going to write in the xml is a reference to the `ObservedData`. It is important though in this case to keep in mind than when deserializing we have to make sure the `ObservedData` has been deserialized before deserializing the corresponding `WeightedObservedFata`, otherwise we might end up with an exception. The observed data of a project is a very good example of this, since they are referenced in many different places of a project. Note for example that if you write a new class that has a `WeightedObservedData` member, when you are serializing it you would also be implicitely keeping a reference to the Observed Data underneath that object - and therefore will have to be carefull that the (de)serialization in your project is correct.
+
+# TypedSerialize(...) - TypedDeserialize(...)
+
+In case the above functionalities do not cover your use case, you can also use `TypedSerialize(TObject objectToSerialize, TContext context)` and `TypedDeserialize(TObject objectToDeserialize, XElement outputToDeserialize, TContext context)` to be able to specify the actions that should happen before and after (de)serialization. This is the functionality used f.e. for the (de)serialization of the FormulaCache and its corresponding StringMap (as discussed in the xml structure section above). 
+
```

Comment:
Also here: would be nice to have a code snippet for TypedSerialize / TypedDeserialize